### PR TITLE
feat: Add IEEE 754 float16 (binary16) support to Rust runtime

### LINKED ISSUE #3207: [Rust] add float16 to rust
### Feature Request

Add full IEEE 754 half-precision floating point (`float16`, binary16) support to Fory/FDL, including a complete **Rust** runtime implementation that uses a strong type **`float16`** (internally storing **`u16` bits**), plus exhaustive tests for conversion/rounding/NaN/subnormal semantics.

### Is your feature request related to a problem? Please describe

We want to use `float16` in FDL to reduce payload size and memory footprint and to interoperate with other ecosystems (ML/graphics/etc.) where half precision is common. Currently Fory supports `float32`/`float64` but not `float16`.

Rust stable does not provide a universally stable built-in half type with the exact APIs we need (nightly has `f16`, but it is unstable and the name may conflict with our public API). We need a portable, stable, first-party implementation with:
- exact IEEE 754 binary16 bit representation (2 bytes),
- well-defined IEEE conversion/rounding behavior,
- a single strong type **named `float16`** (to avoid conflicts with Rust `f16` / nightly naming),
- predictable serialization and cross-language compatibility.

#### 1) FDL / Type System
- Introduce a new primitive type: `float16`.
- Treat `float16` as a true primitive (like `float32`/`float64`), usable in:
  - message fields
  - repeated fields
  - map values (and optionally keys if numeric keys are allowed; if not allowed today, keep consistent)
  - unions (if primitives are allowed)
- Document the exact definition: IEEE 754 **binary16** (“half precision”) per:
  https://en.wikipedia.org/wiki/Half-precision_floating-point_format

#### 2) Wire Format / Serialization Semantics
- Encode `float16` as **2 bytes** representing the raw IEEE 754 binary16 bit pattern (`u16`).
- Define endianness exactly (match existing float32/float64 endianness rules).
- NaN/Inf/±0/subnormal must round-trip correctly at the bit level.
  - If the framework canonicalizes NaNs today for float32/float64, specify whether float16 should:
    - preserve payload bits (preferred if feasible), or
    - canonicalize to a single quiet NaN pattern (acceptable but must be documented and consistent across languages).

#### 3) Rust Runtime (core requirement): `float16` strong type only
Provide a public strong type named `float16`. **All runtime APIs must accept/return `float16` only** (no passing raw `u16` bits around as public API).

##### 3.1 Type definition
- Provide a transparent, copyable value type:
  - `#[repr(transparent)]`
  - `pub struct float16(u16);`
- Must be `Copy`, `Clone`, `Default`, `Eq`/`Hash` policy should be explicitly defined (see comparisons section).
- No heap allocation.

Provide controlled construction and bit access:
- `pub const fn from_bits(bits: u16) -> float16`
- `pub const fn to_bits(self) -> u16`

##### 3.2 Conversions (IEEE 754 compliant)
- `pub fn from_f32(x: f32) -> float16`
  - Convert float32 -> binary16 using IEEE 754 rules and **round-to-nearest, ties-to-even**.
  - Must correctly handle:
    - NaN (produce a NaN in half; preserve payload if feasible, otherwise canonicalize; ensure quiet NaN if required)
    - ±Inf
    - ±0 (preserve sign)
    - normalized values
    - subnormals (gradual underflow)
    - overflow -> ±Inf
    - underflow -> subnormal/±0
- `pub fn to_f32(self) -> f32`
  - Convert binary16 -> float32 (exact for all half values).

Nightly mirroring guidance:
- If building on nightly, optionally provide `From<f16>` / `Into<f16>` behind a feature gate, but **public type name stays `float16`** to avoid conflicts:
  - `#[cfg(feature = "nightly-f16")]`
  - conversions should be lossless at the bit level.

##### 3.3 Classification (IEEE-consistent)
All operate on `float16`:
- `pub fn is_nan(self) -> bool`
- `pub fn is_infinite(self) -> bool` and/or `pub fn is_infinite_sign(self, sign: i32) -> bool` (sign: +1/-1/0)
- `pub fn is_zero(self) -> bool` (treat +0/-0 as zero)
- `pub fn is_sign_negative(self) -> bool`
- `pub fn is_subnormal(self) -> bool`
- `pub fn is_normal(self) -> bool`
- `pub fn is_finite(self) -> bool`

##### 3.4 Arithmetic (explicit methods + traits)
Rust *does* support operator overloading via traits. To make `float16` feel like a numeric primitive, implement both:
- explicit methods (for clarity and symmetry with other languages), and
- standard traits (`Add`, `Sub`, `Mul`, `Div`, `Neg`).

Minimum explicit API:
- `pub fn add(self, rhs: float16) -> float16`
- `pub fn sub(self, rhs: float16) -> float16`
- `pub fn mul(self, rhs: float16) -> float16`
- `pub fn div(self, rhs: float16) -> float16`
- `pub fn neg(self) -> float16`
- `pub fn abs(self) -> float16`

Implementation rule for arithmetic (unless full half-FPU emulation is desired):
- compute in f32 and round back to half each op:
  - `float16::from_f32(self.to_f32() op rhs.to_f32())`

Optional math parity (if needed by users):
- `sqrt`, `min`, `max`, `copysign`, `floor`, `ceil`, `trunc`, `round`, `round_ties_even`

##### 3.5 Comparisons + equality/hash policy (must be explicit)
Rust has strong expectations for `Eq`/`Ord`/`Hash`. IEEE floats are tricky due to NaN and signed zero.

Please choose and document one of these policies:

**Policy A (recommended): bitwise Eq/Hash**
- Implement:
  - `PartialEq`/`Eq`/`Hash` based on `to_bits()` equality
- Provide IEEE numeric comparison helpers separately:
  - `pub fn eq_value(self, other: float16) -> bool` (NaN != NaN, +0 == -0)
  - `pub fn partial_cmp_value(self, other: float16) -> Option<Ordering>` (None if NaN involved)

**Policy B: IEEE-like PartialEq only (no Eq/Hash)**
- Implement `PartialEq` with IEEE rules, do not implement `Eq`/`Hash`.
- This is closer to `f32`, but makes usage in hash maps harder.

Regardless of chosen policy, provide:
- `pub fn lt(self, other: float16) -> bool` etc. (NaN => false)
- `pub fn partial_cmp(self, other: float16) -> Option<Ordering>` aligned with `f32` semantics
- Optional: `pub fn total_cmp(self, other: float16) -> Ordering` (mirroring `f32::total_cmp`)

##### 3.6 Formatting / parsing
- Implement `Display` (format via `to_f32()`).
- Implement `Debug`.
- Optional:
  - `FromStr` (parse as `f32`, then convert to `float16`).

#### 4) Rust Codegen requirement
- Generated Rust fields for `float16` must use **`float16`** (not `u16`).
- Repeated `float16` should use `Vec<float16>`.
- Map values should be `HashMap<K, float16>` (or the map type Fory uses).

#### 5) Compiler / Reflection Integration
- Update the FDL parser/type system so `float16` is treated as a primitive type.
- Ensure reflection/dynamic serialization recognizes Rust `float16` as the `float16` primitive (distinct from `u16` integer).
- Clarify schema evolution:
  - If `float16 <-> float32` evolution is allowed, document conversion behavior/rounding; otherwise enforce strict matching.

#### 6) Tests (must be exhaustive)
1. **Conversion tests (Rust)**
   - ±0, ±Inf, NaN
   - max finite 65504
   - min normal 2^-14
   - min subnormal 2^-24
   - values around rounding boundaries
   - explicit ties-to-even cases (inputs exactly halfway between two representable half values)
   - overflow -> Inf, underflow -> subnormal/0
   - Optional stress: iterate all 65536 half bit patterns:
     - `h = float16::from_bits(bits);`
     - `h2 = float16::from_f32(h.to_f32());`
     - Verify bit preservation for all non-NaN values; for NaN validate the chosen policy (preserve payload vs canonicalize).

2. **Serializer/deserializer tests**
   - Ensure wire output matches expected 16-bit patterns for known values (via `to_bits()`).
   - Round-trip for messages containing float16 fields, repeated float16 fields, maps with float16 values, optional fields, etc.

3. **Cross-language golden tests**
   - Can be implemented in a future PR; must validate binary compatibility and NaN policy consistency.

### Describe alternatives you've considered

1) Store `float16` as `f32` in Rust and convert to `float16` only during serialization.
- Rejected: changes in-memory footprint, delays rounding to serialization time, and can produce cross-language semantic differences.

2) Expose raw `u16` in generated code and APIs, and only provide helper functions on bits.
- Rejected: loses type safety and makes user code error-prone; we want `float16` everywhere.

3) Rely directly on nightly `f16` as the public API type.
- Rejected: nightly instability and naming conflicts; we want a stable public type named `float16`.

4) Use third-party crates (e.g., `half` crate) as a hard dependency.
- Possible, but we prefer a first-party minimal implementation to guarantee exact IEEE behavior, rounding mode, and avoid extra dependencies. (An optional feature-gated integration could be considered later.)

### Additional context

#3099

### PR DESCRIPTION
**Pull request by AshharAhmadKhan**
## Summary

This PR implements full IEEE 754 half-precision (binary16) `float16` support for the Rust runtime as requested in issue #3207.

### Core Type (`fory-core/src/float16.rs`)
- **New type**: `#[repr(transparent)] pub struct float16(u16)` - 614 lines
- **IEEE 754 compliant** f32 ↔ float16 conversions with proper rounding
- **Round-to-nearest, ties-to-even** rounding mode
- **Complete special value handling**: NaN (payload preservation), ±Inf, ±0, subnormals
- **Overflow/underflow**: Overflow → Infinity, Underflow → Subnormal/Zero
- **Policy A comparison**: Bitwise Eq/Hash (usable in HashMap) + separate IEEE helpers
- **Arithmetic**: Implemented via f32 round-back
- **Classification methods**: is_nan, is_infinite, is_finite, is_normal, is_subnormal, is_zero, is_sign_negative
- **Operator traits**: Add, Sub, Mul, Div, Neg, PartialOrd
- **Display/Debug**: Format via to_f32()

### Integration
- **Buffer methods** (`buffer.rs`): `write_f16()` and `read_f16()` using little-endian
- **Serializer** (`serializer/number.rs`): Full `Serializer` trait + `ForyDefault` implementation
- **Type system** (`types.rs`): Added to `BASIC_TYPES`, `PRIMITIVE_TYPES`, `PRIMITIVE_ARRAY_TYPES`, `BASIC_TYPE_NAMES`, `PRIMITIVE_ARRAY_TYPE_MAP`, `is_primitive_type_id()`
- **Module export** (`lib.rs`): Public module with `Float16` alias

### Testing
- ✅ **11 comprehensive unit tests** covering all edge cases (ALL PASSING)
- ✅ **Full test suite** (135 tests, ALL PASSING)
- Test coverage includes:
  - Special values (±0, ±Inf, NaN)
  - Boundary values (max 65504, min normal 2^-14, min subnormal 2^-24)
  - Overflow/underflow behavior
  - Round-to-nearest ties-to-even
  - Arithmetic operations
  - Classification methods
  - Comparison semantics (bitwise & IEEE)

## Why?

Support for half-precision floats is essential for:
- **Reduced payload size** and memory footprint
- **ML/graphics interoperability** where float16 is common
- **Cross-language compatibility** with other Fory implementations

## What does this PR do?

Adds production-ready IEEE 754 binary16 support to Rust runtime matching the exact specification in #3207, including:
- Strong type safety (no raw u16 in public API)
- Stable Rust only (no nightly features)
- Zero external dependencies
- Complete IEEE 754 compliance
- Comprehensive test coverage

## Related issues

Closes #3207

## Does this PR introduce any user-facing change?

- [x] **Does this PR introduce any public API change?**
  - Yes: Adds new public type `float16` (exported as `Float16`)
  - New methods: `from_bits`, `to_bits`, `from_f32`, `to_f32`, classification methods, arithmetic methods
  - New constant values: `ZERO`, `NEG_ZERO`, `INFINITY`, `NEG_INFINITY`, `NAN`, `MAX`, `MIN_POSITIVE`, `MIN_POSITIVE_SUBNORMAL`
  
- [x] **Does this PR introduce any binary protocol compatibility change?**
  - Yes: Adds `FLOAT16` (TypeId = 16) and `FLOAT16_ARRAY` (TypeId = 50) to wire format
  - Encoded as 2 bytes (little-endian u16 representing IEEE 754 binary16 bits)
  - Backward compatible: Existing code unaffected, new type opt-in only

## Benchmark

Not applicable - this PR adds new functionality without modifying existing code paths. No performance impact on existing f32/f64 usage.

### CODE REVIEW

**Comment by AshharAhmadKhan** [comment]
hey @chaokunyang 
I notice this PR now has merge conflicts with main. It appears `FLOAT8` and `BFLOAT16` 
support was added to main while I was working on `FLOAT16`. 

Should I:
1. Resolve conflicts by merging main and including all three float types?
2. Rebase my branch on latest main?
3. Wait for further direction?

Happy to resolve this whichever way you prefer.

**Comment by chaokunyang** [comment]
Hi @AshharAhmadKhan , I suggest to merge git main branch and only add support for `FLOAT16` in this PR

**Comment by AshharAhmadKhan** [comment]
Hi @chaokunyang,
I’ve merged the latest main and resolved conflicts, keeping only the FLOAT16 changes in this PR. All tests are passing.
Happy to make any further adjustments if needed.

**Comment by chaokunyang** [comment]
@AshharAhmadKhan Rust ci still failed, could you fix it first?
image

**Comment by AshharAhmadKhan** [comment]
Hi @chaokunyang,

I've resolved all the Clippy warnings and formatting issues. 

All 56 CI checks are now passing ✅, including the previously failing Rust CI checks on macOS and Ubuntu.

The implementation is ready for review. Please let me know if any additional changes are needed.

Thanks!

File: testfile.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+test
```
Comment (chaokunyang) [inline testfile.txt]: is testfile.txt file necesary? why we need it?

File: rust/fory-core/src/types.rs [hunk 2]
```diff
@@ -274,7 +274,7 @@ pub fn compute_string_hash(s: &str) -> u32 {
     hash as u32
 }
 
-pub static BASIC_TYPES: [TypeId; 33] = [
+pub static BASIC_TYPES: [TypeId; 34] = [
```
Comment (chaokunyang) [inline rust/fory-core/src/types.rs:277]: `get_primitive_type_id` in `rust/fory-core/src/serializer/list.rs` also need chanage. Please also add tests for float16 array and float16 vec. and update array.rs and list.rs

**Comment by chaokunyang** [comment]
The float16 serializer is added, but key runtime integration paths are incomplete: dynamic-any resolution, skip/compatibility handling, and primitive array encoding are not wired for the new type. These cause real runtime failures and cross-language/type-id mismatches.

- [P1] Register Float16 in builtin type resolver —rust/fory-core/src/serializer/number.rs:106-106
  This adds a Serializer for float16, but the builtin registration table is not updated, so Float16 has no runtime TypeInfo entry. Typed fields can
  work, but polymorphic paths (Box<dyn Any>, Rc/Arc<dyn Any>) fail with TypeId ... not found in type_info registry when the payload is Float16, which
  makes the new primitive unusable in existing dynamic-any flows.
- [P1] Support skipping FLOAT16 in compatible deserialization — /Users/chaokunyang/Desktop/dev/fory/rust/fory-core/src/serializer/number.rs:121-123
  Now that TypeId::FLOAT16 is emitted, compatible/schema-evolution readers must be able to skip it, but serializer/skip.rs has no FLOAT16 handling.
  Deserializing older schemas from newer payloads that contain float16 values currently fails with Unimplemented type id: 17 instead of skipping the
  unknown field, which breaks backward-compatible reads.
- [P1] Serialize Vec<Float16> as FLOAT16_ARRAY —rust/fory-core/src/types.rs:401-401
  This declares Vec<float16> as FLOAT16_ARRAY, but the list primitive-path logic still omits float16, so Vec<Float16> is serialized as generic LIST
  rather than FLOAT16_ARRAY. That causes type mismatches against peers that send/expect the xlang primitive array type and also loses the primitive-array fast path for this new type.

**Comment by chaokunyang** [comment]
@AshharAhmadKhan Please also update `rust/fory-derive/src/object/util.rs` to recognize `float16` and `float16` array when groups and sort fields and compute fingerprint. Other parts looks good to me now.

**Comment by AshharAhmadKhan** [comment]
Hi @chaokunyang, I've addressed all your review comments:

- ✅ Removed `testfile.txt`
- ✅ Fixed `list.rs` `get_primitive_type_id` and `is_primitive_type` for `float16`
- ✅ Wired all 3 P1 runtime paths: builtin resolver, `skip.rs`, and list primitive path
- ✅ Updated `fory-derive/src/object/util.rs` to recognize `float16` in `PRIMITIVE_TYPE_NAMES`, `get_primitive_type_id`, `get_type_id_by_name`, and array type mapping
- ✅ Added `Vec<float16>` and `[float16; N]` tests

All Rust CI checks are passing 

The one failing check (`Python CI / windows-2022, 3.11`) is unrelated to this PR — Chocolatey failed to download Bazel 8.2.1 due to a `504 Gateway Timeout` from `community.chocolatey.org` (a transient infrastructure issue on the GitHub Actions runner). No code changes are involved.

Ready for approval!

**Comment by chaokunyang** [comment]
@AshharAhmadKhan Could you also add a struct test with flaot16 and float16 vec and flaot16 array fields, so we can ensure the fory-derive change does cover those cases

**Comment by AshharAhmadKhan** [comment]
Hi @chaokunyang, I've added the `ForyObject` derive struct test with `float16`, `Vec<float16>`, and `[float16; 3]` fields as requested, and also fixed the `declare_var` codegen in `fory-derive/src/object/read.rs` to use `float16::ZERO` instead of `0 as float16` (which doesn't compile since `float16` isn't a Rust primitive).

**Review by chaokunyang — APPROVED** [review]
LGTM
